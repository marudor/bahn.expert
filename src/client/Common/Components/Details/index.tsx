import { format } from 'date-fns';
import { Header } from './Header';
import { StopList } from './StopList';
import { useDetails } from 'client/Common/provider/DetailsProvider';
import { useEffect } from 'react';
import { useHeaderTagsActions } from 'client/Common/provider/HeaderTagProvider';
import type { FC } from 'react';

export const Details: FC = () => {
  const { updateTitle, updateDescription, updateKeywords } =
    useHeaderTagsActions();
  const { initialDepartureDate, details, trainName } = useDetails();

  useEffect(() => {
    if (details) {
      updateTitle(
        `${details.train.name} @ ${format(
          details.departure.time,
          'dd.MM.yyyy',
        )}`,
      );
      updateKeywords([details.train.name]);
    } else {
      updateTitle(trainName);
    }
    let description = `Details zu ${trainName}`;

    if (initialDepartureDate) {
      description += ` am ${format(initialDepartureDate, 'dd.MM.yyyy')}`;
    }
    updateDescription(description);
  }, [
    details,
    initialDepartureDate,
    trainName,
    updateDescription,
    updateTitle,
    updateKeywords,
  ]);

  return (
    <>
      <Header />
      <StopList />
    </>
  );
};
