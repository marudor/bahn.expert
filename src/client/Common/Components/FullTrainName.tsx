import type { CommonProductInfo } from '@/types/HAFAS';
import type { FC } from 'react';

type FTNPropsFallback = {
	train?: CommonProductInfo;
	fallback: string;
};

type FTNProps = {
	train: CommonProductInfo;
	fallback?: string;
};

type Props = FTNProps | FTNPropsFallback;

export const FullTrainName: FC<Props> = ({ train, fallback }) => {
	if (!train) {
		return <span>{fallback}</span>;
	}
	const usesNumberAsIdentification =
		train.number && train.name.endsWith(train.number);
	return (
		<span data-testid="detailsTrainName">
			{train.name}
			{!usesNumberAsIdentification && ` (${train.number})`}
		</span>
	);
};
