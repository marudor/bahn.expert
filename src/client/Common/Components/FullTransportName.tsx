import type { TransportPublicDestination } from '@/external/generated/risConnections';
import type { Transport } from '@/external/generated/risJourneysV2';
import { Link } from '@tanstack/react-router';
import type { FC } from 'react';

interface Props {
	transport:
		| Pick<Transport, 'journeyNumber' | 'line' | 'category'>
		| Pick<TransportPublicDestination, 'category' | 'line' | 'number'>;
	journeyId?: string;
}

export const FullTransportName: FC<Props> = ({ transport, journeyId }) => {
	const journeyNumber =
		'journeyNumber' in transport ? transport.journeyNumber : transport.number;
	let trainName = transport.category;

	if (transport.line) {
		trainName += ` ${transport.line} (${journeyNumber})`;
	} else {
		trainName += ` ${journeyNumber}`;
	}

	if (journeyId) {
		return (
			<Link
				data-testid="fullTransportName"
				to="/details/$train"
				params={{
					train: `${transport.category} ${journeyNumber}`,
				}}
				search={{
					journeyId,
				}}
			>
				{trainName}
			</Link>
		);
	}

	return <span data-testid="fullTransportName">{trainName}</span>;
};
