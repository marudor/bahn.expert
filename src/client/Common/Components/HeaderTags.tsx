import { useHeaderTags } from '@/client/Common/provider/HeaderTagProvider';
import type { FC } from 'react';
import { Meta } from 'react-head';

export const HeaderTags: FC = () => {
	const { title, description, keywords } = useHeaderTags();

	return (
		<>
			<title>{title}</title>
			<Meta name="description" content={description} />
			<Meta name="keywords" content={[...keywords].join(', ')} />
			{/* Open Graph Start - rest in __root.tsx */}
			<Meta property="og:title" content={title} />
			<Meta property="og:description" content={description} />
			{/* Open Graph End */}
		</>
	);
};
