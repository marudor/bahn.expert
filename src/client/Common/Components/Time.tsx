/* eslint no-nested-ternary: 0 */
import { format, subMinutes } from 'date-fns';
import { useCommonConfig } from 'client/Common/provider/CommonConfigProvider';
import styled from '@emotion/styled';
import type { FC } from 'react';

const DelayContainer = styled.span<{ early?: boolean; delayed?: boolean }>(
  ({ theme, early }) => early && theme.mixins.early,
  ({ theme, delayed }) => delayed && theme.mixins.delayed,
);

const Container = styled(DelayContainer.withComponent('div'))<{
  alignEnd?: boolean;
  multiLine?: boolean;
  cancelled?: boolean;
}>(
  {
    display: 'flex',
  },
  ({ alignEnd }) =>
    alignEnd && {
      alignItems: 'flex-end',
    },
  ({ multiLine }) =>
    multiLine && {
      flexDirection: 'column',
    },
  ({ theme, cancelled }) => cancelled && theme.mixins.cancelled,
);

const TimeContainer = styled.span<{ oneLine?: boolean }>(
  ({ oneLine }) =>
    oneLine && {
      marginRight: '.2em',
    },
);

interface Props {
  alignEnd?: boolean;
  className?: string;
  delay?: number;
  real?: Date;
  showZero?: boolean;
  oneLine?: boolean;
  cancelled?: boolean;
}

function delayString(delay: number) {
  if (delay < 0) {
    return `-${Math.abs(delay)}`;
  }

  return `+${delay}`;
}

export const Time: FC<Props> = ({
  className,
  delay,
  real,
  showZero = true,
  alignEnd,
  oneLine,
  cancelled,
}) => {
  const showOriginalTime = !useCommonConfig().time;

  if (!real) return null;
  const time = showOriginalTime && delay ? subMinutes(real, delay) : real;

  const hasDelay = showZero ? delay != null : Boolean(delay);

  return (
    <Container
      data-testid="timeContainer"
      className={className}
      delayed={Boolean(!showOriginalTime && hasDelay && delay && delay > 0)}
      early={Boolean(
        !showOriginalTime && hasDelay && delay != null && delay <= 0,
      )}
      alignEnd={alignEnd}
      multiLine={!oneLine}
      cancelled={cancelled}
    >
      <TimeContainer oneLine={oneLine} data-testid="time">
        {format(time, 'HH:mm')}
      </TimeContainer>
      {hasDelay && (
        <DelayContainer
          delayed={Boolean(delay && delay > 0)}
          early={Boolean(delay && delay <= 0)}
          data-testid="delay"
        >
          {delayString(delay as number)}
        </DelayContainer>
      )}
    </Container>
  );
};
