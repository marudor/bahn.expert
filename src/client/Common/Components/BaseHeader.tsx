import { AppBar, IconButton, styled, Toolbar } from '@mui/material';
import { HeaderTags } from './HeaderTags';
import { Menu } from '@mui/icons-material';
import { NavigationContext } from '@/client/Common/Components/Navigation/NavigationContext';
import { useContext } from 'react';
import type { FC, PropsWithChildren } from 'react';

const NotchFix = styled('div')(({ theme }) => ({
  top: '-5em',
  height: '5em',
  width: '100%',
  position: 'fixed',
  zIndex: 99999,
  background: theme.vars.palette.background.default,
}));

const Offset = styled('div')<{ spacing?: number }>(
  ({ theme, spacing = 0 }) => ({
    minHeight: `calc(${theme.vars.shape.headerSpacing} + ${spacing}em)`,
  }),
);

interface Props {
  /** Unit: em */
  spacing?: number;
}

export const BaseHeader: FC<PropsWithChildren<Props>> = ({
  children,
  spacing,
}) => {
  const { toggleDrawer } = useContext(NavigationContext);

  return (
    <>
      <NotchFix />
      <AppBar enableColorOnDark position="fixed" data-testid="header">
        <HeaderTags />
        <Toolbar disableGutters>
          <IconButton
            size="small"
            data-testid="navToggle"
            aria-label="Menu"
            onClick={toggleDrawer}
            color="inherit"
          >
            <Menu color="inherit" />
          </IconButton>
          {children}
        </Toolbar>
      </AppBar>
      <Offset spacing={spacing} />
    </>
  );
};
