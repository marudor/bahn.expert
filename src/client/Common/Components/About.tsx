import { BaseHeader } from '@/client/Common/Components/BaseHeader';
import { Stack } from '@mui/material';
import type { FC } from 'react';

const Privacy = () => (
  <div data-testid="Privacy">
    <h2>Impressum</h2>
    <p>Verantwortlich für diese Seite ist:</p>
    <p>
      <span>
        {globalThis.IMPRINT.name}
        <br />
        {globalThis.IMPRINT.street}
        <br />
        {globalThis.IMPRINT.town}
      </span>
    </p>
    <h2>Datenschutzerklärung</h2>
    <h3>Logdaten des Webservers</h3>
    <p>
      Bei jedem Aufruf dieser Website werden vom Webserver die folgenden Daten
      gespeichert und für vier Wochen aufbewahrt:
    </p>
    <p>
      Datum und Uhrzeit,
      <br /> die aufgerufene Seite,
      <br /> der dabei verwendete Browser (gemäß des von diesem gesendeten
      UserAgents) und
      <br /> der Referer (d.h. die auf diese Seite verlinkende Website —
      ebenfalls gemäß der vom Browser gesendeten Informationen, kann meist im
      Browser deaktiviert werden).
      <br />
      Die IP-Adresse sowie die bei Aufruf der Hauptseite an den Server
      übertragenen Geokoordinaten werden nicht gespeichert. Eine Ausnahme für
      das Speichern der IP ist die Nutzung von deprecated APIs. In diesen Fällen
      wird die IP gespeichert bis die API entfernt wird um eventuelle Nutzern
      bescheid zu geben.
    </p>
  </div>
);

export const About: FC = () => {
  return (
    <>
      <BaseHeader spacing={0.3}>About</BaseHeader>
      <Stack margin="0 10px 0">
        <span>
          Entwickelt von{' '}
          <a
            href="https://chaos.social/@marudor"
            target="_blank"
            rel="noopener noreferrer"
          >
            @marudor@chaos.social
          </a>
        </span>
        <span>
          Featurewünsche und Bugreports sind mir am liebsten über Github. Dafür
          gibt es jeweils ein eigenes Formular.{' '}
        </span>
        <span>
          <a
            href="https://github.com/marudor/bahn.expert/issues/new?assignees=&labels=bug&template=bug_report.md&title="
            target="_blank"
            rel="noopener noreferrer"
          >
            Bugs
          </a>{' '}
          sowie{' '}
          <a
            href="https://github.com/marudor/bahn.expert/issues/new?assignees=&labels=feature&template=feature_request.md&title="
            target="_blank"
            rel="noopener noreferrer"
          >
            Features
          </a>
        </span>
        <span>
          Wer mich anderweitig kontaktieren möchte kann dies neben dem
          verlinkten Social Media Profil auch per mail an info (at) bahn (dot)
          expert. Bitte als Anrede nicht siezen. Im Idealfall einfach "marudor"
          nutzen.
        </span>
        <Privacy />
      </Stack>
    </>
  );
};
// eslint-disable-next-line import/no-default-export
export default About;
