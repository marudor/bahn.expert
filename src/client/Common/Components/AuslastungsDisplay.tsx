import { SingleAuslastungsDisplay } from '@/client/Common/Components/SingleAuslastungsDisplay';
import { Stack, styled } from '@mui/material';
import type { ComponentProps, FC } from 'react';
import type { RouteAuslastung } from '@/types/routing';

const Container = styled('div')<{ oneLine?: boolean }>(
  {
    fontSize: '.75em',
    display: 'flex',
  },
  ({ oneLine }) =>
    !oneLine && {
      flexDirection: 'column',
    },
);

const Seperator = styled('span')`
  margin: 0 0.25em;
`;

export interface Props extends ComponentProps<'div'> {
  auslastung: RouteAuslastung;
  oneLine?: boolean;
}

export const AuslastungsDisplay: FC<Props> = ({
  auslastung,
  oneLine,
  ...rest
}) => {
  if (!auslastung.first && !auslastung.second) {
    return null;
  }
  return (
    <Container oneLine={oneLine} data-testid="auslastungDisplay" {...rest}>
      Auslastung
      <Stack direction="row" marginLeft=".2em" component="span">
        <div data-testid="first">
          1. <SingleAuslastungsDisplay auslastung={auslastung.first} />
        </div>
        <Seperator>|</Seperator>
        <div data-testid="second">
          2. <SingleAuslastungsDisplay auslastung={auslastung.second} />
        </div>
      </Stack>
    </Container>
  );
};
