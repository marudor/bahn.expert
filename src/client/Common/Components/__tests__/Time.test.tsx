import { render } from '@/client/__tests__/testHelper';
import { screen } from '@testing-library/react';
import { Time } from '@/client/Common/Components/Time';

const sampleTime = new Date('2019-06-12T13:55:37.648Z');

describe('Time', () => {
  it('no time', () => {
    const { container } = render(<Time />);

    expect(container).toBeNull();
  });
  it('only scheduled Data', () => {
    render(<Time real={sampleTime} />);

    expect(screen.queryByTestId('realTimeOrDelay')).toBeNull();
    expect(screen.queryByTestId('timeToDisplay')).toHaveTextContent('13:55');
  });

  it('10 Minutes delay', () => {
    const { container, theme } = render(<Time real={sampleTime} delay={10} />);

    expect(screen.queryByTestId('timeToDisplay')).toHaveTextContent('13:45');
    expect(screen.queryByTestId('realTimeOrDelay')).toHaveTextContent('13:55');
    expect(screen.queryByTestId('realTimeOrDelay')).toHaveStyle(
      `color: ${theme.colors.red}`,
    );
    expect(container).toMatchInlineSnapshot(`
      .emotion-1 {
        display: -webkit-box;
        display: -webkit-flex;
        display: -ms-flexbox;
        display: flex;
        -webkit-flex-direction: column;
        -ms-flex-direction: column;
        flex-direction: column;
        font-size: 0.9em;
      }

      .emotion-2 {
        margin-right: .2em;
      }

      .emotion-4 {
        margin-right: .2em;
        color: #ff1744;
      }

      <div
        class="MuiStack-root emotion-0 emotion-1"
        data-testid="timeContainer"
      >
        <span
          class="emotion-2 emotion-3"
          data-testid="timeToDisplay"
        >
          13:45
        </span>
        <span
          class="emotion-4 emotion-3"
          data-testid="realTimeOrDelay"
        >
          13:55
        </span>
      </div>
    `);
  });

  it('5 Minutes early', () => {
    const { container, theme } = render(<Time real={sampleTime} delay={-5} />);

    expect(screen.queryByTestId('timeToDisplay')).toHaveTextContent('14:00');
    expect(screen.queryByTestId('realTimeOrDelay')).toHaveTextContent('13:55');
    expect(screen.queryByTestId('realTimeOrDelay')).toHaveStyle(
      `color: ${theme.colors.green}`,
    );
    expect(container).toMatchInlineSnapshot(`
      .emotion-1 {
        display: -webkit-box;
        display: -webkit-flex;
        display: -ms-flexbox;
        display: flex;
        -webkit-flex-direction: column;
        -ms-flex-direction: column;
        flex-direction: column;
        font-size: 0.9em;
      }

      .emotion-2 {
        margin-right: .2em;
      }

      .emotion-4 {
        margin-right: .2em;
        color: #7cb342;
      }

      <div
        class="MuiStack-root emotion-0 emotion-1"
        data-testid="timeContainer"
      >
        <span
          class="emotion-2 emotion-3"
          data-testid="timeToDisplay"
        >
          14:00
        </span>
        <span
          class="emotion-4 emotion-3"
          data-testid="realTimeOrDelay"
        >
          13:55
        </span>
      </div>
    `);
  });

  it('shows 0 delay number', () => {
    const { container, theme } = render(<Time real={sampleTime} delay={0} />);

    expect(screen.queryByTestId('timeToDisplay')).toHaveTextContent('13:55');
    expect(screen.queryByTestId('realTimeOrDelay')).toHaveTextContent('13:55');
    expect(screen.queryByTestId('realTimeOrDelay')).toHaveStyle(
      `color: ${theme.colors.green}`,
    );
    expect(container).toMatchInlineSnapshot(`
      .emotion-1 {
        display: -webkit-box;
        display: -webkit-flex;
        display: -ms-flexbox;
        display: flex;
        -webkit-flex-direction: column;
        -ms-flex-direction: column;
        flex-direction: column;
        font-size: 0.9em;
      }

      .emotion-2 {
        margin-right: .2em;
      }

      .emotion-4 {
        margin-right: .2em;
        color: #7cb342;
      }

      <div
        class="MuiStack-root emotion-0 emotion-1"
        data-testid="timeContainer"
      >
        <span
          class="emotion-2 emotion-3"
          data-testid="timeToDisplay"
        >
          13:55
        </span>
        <span
          class="emotion-4 emotion-3"
          data-testid="realTimeOrDelay"
        >
          13:55
        </span>
      </div>
    `);
  });
});
