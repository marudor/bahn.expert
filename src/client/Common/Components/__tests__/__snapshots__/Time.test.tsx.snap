// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Time does not show 0 delay number 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="emotion-0"
  data-testid="timeContainer"
>
  <span
    class="emotion-1"
    data-testid="time"
  >
    13:55
  </span>
</div>
`;

exports[`Time no time 1`] = `null`;

exports[`Time showOriginalTime = false 5 Minutes early 1`] = `
.emotion-0 {
  color: #7cb342;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2 {
  color: #7cb342;
}

<div
  class="emotion-0"
  data-testid="timeContainer"
>
  <span
    class="emotion-1"
    data-testid="time"
  >
    13:55
  </span>
  <span
    class="emotion-2"
    data-testid="delay"
  >
    -5
  </span>
</div>
`;

exports[`Time showOriginalTime = false 10 Minutes delay 1`] = `
.emotion-0 {
  color: #ff1744;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2 {
  color: #ff1744;
}

<div
  class="emotion-0"
  data-testid="timeContainer"
>
  <span
    class="emotion-1"
    data-testid="time"
  >
    13:55
  </span>
  <span
    class="emotion-2"
    data-testid="delay"
  >
    +10
  </span>
</div>
`;

exports[`Time showOriginalTime = true 5 Minutes early 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2 {
  color: #7cb342;
}

.emotion-2 {
  color: #7cb342;
}

<div
  class="emotion-0"
  data-testid="timeContainer"
>
  <span
    class="emotion-1"
    data-testid="time"
  >
    14:00
  </span>
  <span
    class="emotion-2"
    data-testid="delay"
  >
    -5
  </span>
</div>
`;

exports[`Time showOriginalTime = true 10 Minutes delay 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-2 {
  color: #ff1744;
}

.emotion-2 {
  color: #ff1744;
}

<div
  class="emotion-0"
  data-testid="timeContainer"
>
  <span
    class="emotion-1"
    data-testid="time"
  >
    13:45
  </span>
  <span
    class="emotion-2"
    data-testid="delay"
  >
    +10
  </span>
</div>
`;

exports[`Time shows 0 delay number 1`] = `
.emotion-0 {
  color: #7cb342;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-0 {
  color: #7cb342;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="emotion-0"
  data-testid="timeContainer"
>
  <span
    class="emotion-1"
    data-testid="time"
  >
    13:55
  </span>
  <span
    class="emotion-1"
    data-testid="delay"
  >
    +0
  </span>
</div>
`;
